core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java

```diff
@@ -212,17 +212,26 @@ public final Object visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEnt
 
    private Object visitDataReadCommand(InvocationContext ctx, AbstractDataCommand command) {
       final Object key = command.getKey();
-      entryFactory.wrapEntryForReading(ctx, key, command.getSegment(), ignoreOwnership(command) || canRead(command));
-      return invokeNextThenApply(ctx, command, dataReadReturnHandler);
+      CompletionStage<Void> stage = entryFactory.wrapEntryForReading(ctx, key, command.getSegment(),
+            ignoreOwnership(command) || canRead(command));
+      if (CompletionStages.isCompletedSuccessfully(stage)) {
+         return invokeNextThenApply(ctx, command, dataReadReturnHandler);
+      }
+      return asyncValue(stage).thenApply(ctx, command, ((rCtx, rCommand, rv) ->
+         invokeNextThenApply(rCtx, rCommand, dataReadReturnHandler)));
    }
 
    @Override
    public Object visitGetAllCommand(InvocationContext ctx, GetAllCommand command) throws Throwable {
       boolean ignoreOwnership = ignoreOwnership(command);
+      AggregateCompletionStage<Void> aggregateCompletionStage = null;
       for (Object key : command.getKeys()) {
-         entryFactory.wrapEntryForReading(ctx, key, keyPartitioner.getSegment(key), ignoreOwnership || canReadKey(key));
+         CompletionStage<Void> stage = entryFactory.wrapEntryForReading(ctx, key, keyPartitioner.getSegment(key),
+               ignoreOwnership || canReadKey(key));
+         aggregateCompletionStage = accumulateStage(stage, aggregateCompletionStage);
       }
-      return invokeNextAndHandle(ctx, command, (rCtx, getAllCommand, rv, t) -> {
+
+      InvocationFinallyFunction<GetAllCommand> handleFunction = (rCtx, getAllCommand, rv, t) -> {
```

Comment:
offtopic: can be cached as well :) 
ps. I'm sounding like Dan xD

Reply:
I cached this one as it is used in both code paths.